### ISSUE TITLE
Suggest `and_then`

### ISSUE THREAD

**Issue by Alexhuszagh** [rust-lang/rust-clippy#9003]
### What it does

If there is a callback that takes a type `T` and returns `Option<U>` (or `Cb: Fn(T) -> Option<U>`), if we have type `Option<T>`, we should prefer `opt_t.and_then(cb)`, and not `opt_t.map(cb)`. This means the result will be `Option<U>` and not `Option<Option<U>>`.

### Lint Name

option_map_option

### Category

style

### Advantage

It avoids nested pattern matching to resolve the nested options and/or results.

### Drawbacks

Although unlikely, if there is an API that expected a nested option/result, then this could produce a spurious warning.

### Example

```rust
pub fn map<T, U: Default, Cb: Fn(T) -> Option<U>>(opt: Option<T>, cb: Cb) -> U {
    if let Some(u) = opt.map(cb) {
        u.unwrap_or_default()
    } else {
        U::default()
    }
}
```

Could be written as:

```rust
pub fn and_then<T, U: Default, Cb: Fn(T) -> Option<U>>(opt: Option<T>, cb: Cb) -> U {
    opt.and_then(cb).unwrap_or_default()
}
```

The warning would be something like:

```
if let Some(u) = opt.map(cb) {
^^^^^^^^^^^ help: try this: `let u = opt.and_then(cb);`
```

**Comment by Emilgardis** [comment]
Specifically, I think this makes the most sense only when used in pattern matching.

We already have the `pedantic` lint `option_option` which should catch if the type is used used in a typedef or fn signature

```rust
fn some_fn<T, F>(x: T) -> Option<F> {todo!()}

if let Some(Some(x)) = x.map(|x| some_fn(x)) {
    todo!()
}
```

**Comment by mikerite** [comment]
`map` and `and_then` don't do the same thing. This lint _only_ makes sense if the return from `map` is immediately unwrapped somehow.
